Fixes #8979 - Jetty 12 - HttpClientTransport network "modes".

sbordet opened the pull request:
* Introduced `oej.io.TransportProtocol` as the abstraction for the low-level transport of high-level protocols. Now protocols such as HTTP/1.1 or HTTP/2 can be transported over QUIC, Unix-Domain, memory, and possibly over other low-level custom protocols too.
* Introduced `oej.client.Request.transportProtocol(TransportProtocol)` to specify `TransportProtocol` for each request.
* Introduced `TransportProtocol` to `[HTTP2Client|HTTP3Client].connect(...)` methods.
* Introduced `[Client|Server]QuicConfiguration` so that it can be used in other Connectors such as `MemoryConnector`.
* Introduced `oej.server.MemoryConnector` and `EndPoint.Pipe` for memory communication between peers, along with a `MemoryTransportProtocol`.
* Introduced `QuicTransportProtocol` as a wrapper for other `TransportProtocol`s, so that QUIC can now also be transported over memory.

sbordet committed:
Fixes #8979 - Jetty 12 - HttpClientTransport network "modes".

* Introduced oej.io.TransportProtocol as the abstraction for the low-level transport of high-level protocols.
Now protocols such as HTTP/1.1 or HTTP/2 can be transported over QUIC, Unix-Domain, memory, and possibly over other low-level custom protocols too.
* Introduced oej.client.Request.transportProtocol(TransportProtocol) to specify TransportProtocol for each request.
* Introduced TransportProtocol to [HTTP2Client|HTTP3Client].connect(...) methods.
* Introduced [Client|Server]QuicConfiguration so that it can be used in other Connectors such as MemoryConnector.
* Introduced oej.server.MemoryConnector and EndPoint.Pipe for memory communication between peers, along with a MemoryTransportProtocol.
* Introduced QuicTransportProtocol as a wrapper for other TransportProtocols, so that QUIC can now also be transported over memory.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
* Improved javadocs and documentation.
* Removed usage of ClientConnector.forUnixDomain() from FastCGIProxyServlet (ee10 and ee9).
* Replaced usage of HTTP3ServerConnector with QuicServerConnector in jetty-http3.xml.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixed invocation to HTTP3ServerConnectionFactory constructor in jetty-http3.xml.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
* Added documentation about TransportProtocol.
* Updated client documentation to consider TransportProtocol.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -211,6 +216,24 @@ default int fill(ByteBuffer buffer) throws IOException
+     * @return the peer address that sent the data, or {@link #EOF}
+     * @throws IOException if the receive fails
+     */
+    default SocketAddress receive(ByteBuffer buffer) throws IOException
```
Is this blocking?

gregw replied:
hmmm few of the existing methods have javadoc making it clear their blocking behaviour.   We should fix this .

gregw commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/MemoryEndPointPipe.java:
```diff
@@ -0,0 +1,344 @@
+/**
+ * <p>Memory-based implementation of {@link EndPoint.Pipe}.</p>
+ */
+public class MemoryEndPointPipe implements EndPoint.Pipe
```
Should the LocalConnector be changed to use this?

sbordet replied:
It is difficult, as `LocalEndPoint` is somehow incompatible with `MemoryEndPoint`.
We did not document `LocalConnector`, which we can keep as testing facility (although in `src/main`), and `MemoryConnector` is now documented.

sbordet committed:
Improved javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

marchof commented:
@sbordet Are there any instructions how to build this locally for testing? Local Maven build fails with the same errors than the CI.

sbordet commented:
@marchof build locally with `mvn clean install -DskipTests`.

The CI failure is due to javadocs, I'm fixing it right now.

sbordet committed:
Fixed javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

marchof commented:
@sbordet Thanks for this effort! Finally I got a super simple prototype running with a naive `TransportProtocol`and `EndPoint` implementation. I have to admit that my brain can hardly handle the abstractions used in these interfaces. But in the end I only needed to implement these methods (probably missing several corner cases):

* `TransportProtocol.connect()`
* `EndPoint.fillInterested()`
* `EndPoint.fill()`
* `Endpoint.write()`

sbordet commented:
@marchof you should not be needing to implement a new `TransportProtocol` and `EndPoint`.

If you explain your use case in more details, perhaps we can help you save some unnecessary work.

marchof commented:
@sbordet Let me try to explain my use case: A proxy server which proxies incoming HTTPS requests over SSH tunnels to HTTP servers. The SSH server managing the tunnels itself is implemented in Java with Apache Mina.

So if implemented in Jetty I see the following flow:

* A Jetty server accepts HTTPS requests and handles it with a `ProxyServlet`.
* The `ProxyServlet` uses a HttpClient with a custom transport
* The custom transport delegates to the Apache Mina API for ssh tunnels, which is basically writing/receiving `byte[]` chunks

sbordet commented:
@marchof let me try to write a test case similar to your use case. I'll ping you when done.

sbordet commented:
@marchof I need more information.

Are you using SSH port forwarding? 

If so, then this is what I think it happens:
* Mina sets up port forwarding; now you have `localhost:1234` on your proxy server that forwards to the remote server.
* A request arrives to the Servlet
* You use `HttpClient` to send a request to `localhost:1234` with the Servlet request information and you're done.

Does Mina allow you to bypass the connection to `localhost:1234`?
Can you give Mina just a `ByteBuffer`, as if you have written it to `localhost:1234` using a socket, therefore bypassing the whole network stack?

marchof commented:
@sbordet Yes, I'm using SSH port forwarding but without using sockets on the server side. This avoids the network stack and allows me to separate the forwards of different users. The forwarding between http hosts and ssh sockets is pure virtual and happens in memory. Therefore I cannot use a HTTP client which connects to a (local) TCP port.

 Apache Mina supports this, you can basically implement your own forwarding endpoints. The use something similar to `ByteBuffer` called `org.apache.sshd.common.util.buffer.Buffer`, but in the end it's just passing around `byte[]`instances.

sbordet commented:
@marchof thanks, we'll write a test case for this scenario.

sbordet committed:
Fixed handling of Instruction notifications in case of re-entrance.
Now first clear the list, then notify to avoid that when re-entering the same instruction is notified multiple times.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Added test case for custom TransportProtocol.
Introduced ContentSourceRequestContent, and updated ProxyHandler to use it.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet commented:
@marchof see `CustomTransportProtocolTest` in this PR. Should be almost identical to your use case.

marchof commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/CustomTransportProtocolTest.java:
```diff
@@ -0,0 +1,395 @@
+                    {
+                        // Copy the response headers.
+                        response.getHeaders().add(r.getHeaders());
+                        // Remote Content-Encoding, as the content has already been decoded.
```
```suggestion
                        // Remove Content-Encoding, as the content has already been decoded.
```

sbordet committed:
Refactored TransportProtocol tests.
Now unix-domain files are temp files, so their name is unique.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

cowwoc commented on documentation/jetty-documentation/src/main/java/org/eclipse/jetty/docs/programming/client/http/HTTPClientDocs.java:
```diff
@@ -919,57 +929,134 @@ public void dynamicOneProtocol()
+        HTTP3Client http3Client = new HTTP3Client(quicConfiguration, connector);
+        ClientConnectionFactoryOverHTTP3.HTTP3 http3 = new ClientConnectionFactoryOverHTTP3.HTTP3(http3Client);
+
+        // The order of the protocols indicates the client's preference.
```
Can you please clarify (in the documentation and test code) whether the order is from the most preferred to least preferred, or the other way around? In the following line, does the client prefer HTTP/1 first, HTTP/2 second and HTTP/3 third?

Is it technically possible to flip that around and have it try HTTP/3, followed by HTTP/2, followed by HTTP/1? Or is this not possible because the connection type needs to be upgraded through the earlier versions?

sbordet replied:
@cowwoc clarified the preference.

It is possible to reverse the order. 

However, since HTTP/3 is not compatible with HTTP/2 and HTTP/1, there is no way to "downgrade" the attempt (and switch from UDP to TCP), so if the server does not support HTTP/3 the request will fail.
Applications may handle the failure and attempt to specify explicitly the HTTP version via `request.version(...)`.
This is not done automatically by `HttpClient`.

The reverse order may be useful if you want your default to be HTTP/3, but you want to explicitly specify HTTP/2 or HTTP/1 for those servers you know don't support HTTP/3.

cowwoc replied:
@sbordet if the order of preference is HTTP/1, HTTP/2, HTTP/3 does that mean that the client will settle on HTTP/1 if the server supports all three protocols? Or will it try to upgrade the connection to HTTP/3?

I suspect that the client will settle on HTTP/1. If this is the case, how could one configure the client to try HTTP/3 and if that fails try HTTP/1 followed by an upgrade to HTTP/2? And if the upgrade to HTTP/2 fails then stay on HTTP/1.

I may be wrong, but this sounds like the happy path to me.

sbordet replied:
HTTP/2 does not usually work with upgrades, it's negotiated via TLS + ALPN.

A client configured with `http/1.1, h2` will send that to the server, and the server will likely pick `h2` because it's the server that decides.

Trying HTTP/3 with a server that does not support it will result in a failure after the connect timeout (differently from TCP, where the failure is almost immediate).
After that timeout, the application may try explicitly HTTP/2.

That's why it's probably most common to do the opposite: if you know the server supports HTTP/3, you explicitly try that; otherwise you are not explicit and therefore HTTP/1.1+HTTP/2 will be negotiated.

Likely, as of now, the best configuration for the client is therefore: HTTP/2, HTTP/1.1, HTTP/3.

cowwoc replied:
As usual, nothing is simple in life :) Check out this interesting discussion related to HTTP/3 negotiation: https://stackoverflow.com/a/77408325/14731 and https://stackoverflow.com/a/67470164/14731

I assume that if support is added for these approaches then users can move from the order HTTP/2, HTTP/1.1, HTTP/3 to HTTP/3, HTTP/2, HTTP/1.1.

I'm not sure what Jetty can do with the `Alt-Svc` approach though. In the case of browsers, they say that future connections will then try HTTP/3 first but in the case of Jetty can the `HttpClient` instance do the same? Meaning, the first connection would try HTTP/2, HTTP/1.1, HTTP/3 but if HTTP/3 support is found then subsequent connections to the same host would try HTTP/3 first.

sbordet replied:
We can't use the DNS extension as it is not available from Java.

About `Alt-Svc`, you have to think of `HttpClient` as `curl`: I doubt that if you make a request with curl saying "use HTTP/2", then curl will store the fact that there was an `Alt-Svc` header and the next request will be done via HTTP/3.

Applications can do that logic, because I guess whatever we do it will break someone.

For example, we do not want to do "preflight" requests to figure out `Alt-Svc` under the covers, as people will expect `HttpClient` to do exactly the request that they are telling it to do, not more.
Some other people will be happy if we do, and we don't want to have an option for all possible combinations of "how to discover HTTP/3".
Some other people won't care about HTTP/3, and really want to use HTTP/1.1 or HTTP/2.
A user can configure the dynamic transport with preference for HTTP/1.1, so it really wants to use HTTP/1.1, even if the server supports HTTP/3.
Just because HTTP/3 is the latest spec, does not mean people want to use it: it is not implicitly better than HTTP/1.1 in this context (it may be for browsers).

`HttpClient` is lower level than browsers, whose job is to show pages no matter how they are retrieved.

@cowwoc I am not sure whether you are asking for a feature here, or just discussing ideas.
I'm happy to discuss either.

sbordet replied:
> Meaning, the first connection would try HTTP/2, HTTP/1.1, HTTP/3 but if HTTP/3 support is found then subsequent connections to the same host would try HTTP/3 first.

That would be possible, I guess. Need to parse the `Alt-Svc` header or the [`ALTSVC` frame](https://datatracker.ietf.org/doc/html/rfc7838#section-7.2), and then the `HttpClientTransportDynamic` could match scheme+authority on the destinations when the HTTP version is not explicitly set in the request, and pull out the preferred `Destination` to use.

As I said, that implies an order relationship based on the protocol (`HTTP/3.compareTo(HTTP/2) > 0`) that may be true for browsers, but not for `HttpClient`.

@cowwoc if you really want this, please file a different issue, because it is out of scope for this PR. Thanks!

cowwoc replied:
@sbordet I guess the only thing I'm wondering is whether:

1. I can start out with an HTTP/2, HTTP/1.1, HTTP/3 preference.
2. Upon making the first call to a hostname, take note whether the server response contains a `Alt-Svc: h3="<port number>` header.
3. Configure Jetty to use a different order of preference for that host in future calls.

If Jetty doesn't provide a way to set preferences on a per-host basis, one thing I could do is create one `HttpClient` per preference order (one normal, one that prefers HTTP/3) and use application-level logic to issue requests with one or the other depending on whether the host is known to support HTTP/3.

sbordet replied:
Bullet 3 should be: Explicitly set the HTTP version on the request via `request.version(HttpVersion.HTTP_3)`.

That's it.

The option of configuring different `HttpClient` instances is not one I would suggest.

cowwoc replied:
@sbordet Filed https://github.com/jetty/jetty.project/issues/11407. Thank you.

cowwoc replied:
I got HTTPS/3 working on the server side but I can't get past the dreaded "PKIX path building failed" exception on the client side. Even hitting "google.com" fails with this error. I'll let you know once I figure this out.

cowwoc replied:
@sbordet It looks like this PR is fine. See https://github.com/jetty/jetty.project/issues/11417 for a follow-up discussion.

sbordet committed:
Clarified protocol preference.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12/8979/httpclient-network-modes'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Renamed `TransportProtocol` to `Transport`.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

lorban commented on jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/EndPoint.java:
```diff
@@ -439,4 +473,20 @@ static SslSessionData withSslSessionId(SslSessionData baseData, String sslSessio
+    /**
+     * <p>A communication conduit between two peers.</p>
+     */
+    interface Pipe
```
I'm not sure we need this interface in the IO API.

lorban reviewed (dismissed):
Besides `EndPoint.Pipe` which I think isn't needed as I do not think we're going to add other pipe connectors in the future, LGTM.

sbordet committed:
Improved documentation.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Merged branch 'jetty-12.0.x' into 'fix/jetty-12/8979/httpclient-network-modes'.

sbordet committed:
Merged branch '12.0.x' into 'fix/jetty-12/8979/httpclient-network-modes'.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Renamed Transport.requiresDomainNamesResolution() -> requiresDomainNameResolution().
Marked HttpDestination constructor as deprecated for removal.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
